buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+
+        final Map<Integer, List<Version>> groupByMinor;
+        if (currentMajorVersions.size() == 1) {
+            // Current is an unreleased major: x.0.0 so we have to look for other unreleased versions in the previous major
+            groupByMinor = previousMajorVersions.stream()
+                .collect(Collectors.groupingBy(Version::getMinor, Collectors.toList()));
+        } else {
+            groupByMinor = currentMajorVersions.stream()
+                .collect(Collectors.groupingBy(Version::getMinor, Collectors.toList()));
+        }
+        return groupByMinor;
+    }
+
+    public void compareToAuthoritative(List<Version> authoritativeReleasedVersions) {
+        Set<Version> notReallyReleased = new HashSet<>(getReleased());
+        notReallyReleased.removeAll(authoritativeReleasedVersions);
+        if (notReallyReleased.isEmpty() == false) {
+            throw new IllegalStateException(
+                "out-of-date released versions" +
+                    "\nFollowing versions are not really released, but the build thinks they are: " + notReallyReleased
+            );
+        }
+
+        Set<Version> incorrectlyConsideredUnreleased = new HashSet<>(authoritativeReleasedVersions);
+        incorrectlyConsideredUnreleased.retainAll(getUnreleased());
+        if (incorrectlyConsideredUnreleased.isEmpty() == false) {
+            throw new IllegalStateException(
+                "out-of-date released versions" +
+                    "\nBuild considers versions unreleased, " +
+                    "but they are released according to an authoritative source: " + incorrectlyConsideredUnreleased +
+                    "\nThe next versions probably needs to be added to Version.java (CURRENT doesn't count)."
+            );
+        }
+    }
+
+    private List<Version> getReleased() {
+        List<Version> unreleased = getUnreleased();
+        return groupByMajor.values().stream()
+            .flatMap(Collection::stream)
+            .filter(each -> unreleased.contains(each) == false)
+            .sorted()
+            .collect(Collectors.toList());
+    }
+
+    public List<Version> getIndexCompatible() {
+        return unmodifiableList(
+            groupByMajor.values().stream()
+                .flatMap(Collection::stream)
+                .filter(version -> version.equals(currentVersion) == false)
+                .sorted()
+                .collect(Collectors.toList())
+        );
+    }
+
+    public List<Version> getWireCompatible() {
+        Version lastPreviousMinor = getLatestVersionByKey(groupByMajor, currentVersion.getMajor() - 1);
+        Version firstPreviousMinor = new Version(lastPreviousMinor.getMajor(), lastPreviousMinor.getMinor(), 0);
+        return unmodifiableList(
+            groupByMajor.values()
+                .stream()
+                .flatMap(Collection::stream)
+                .filter(version -> version.compareTo(firstPreviousMinor) >= 0)
```

Comment:
I think this can be done more simply without streams. Wire compatible is all versions in the current major, and then iterate from the end of the list of versions for the previous major, and stop when the minor version changes.